operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java

```diff
@@ -18,59 +18,101 @@
+        SharedIndexInformer<Deployment> deploymentInformer =
+                client.apps().deployments().inAnyNamespace()
+                        .withLabels(Constants.DEFAULT_LABELS)
+                        .runnableInformer(0);
+
+        EventSource deploymentEvent = new InformerEventSource<>(
+                deploymentInformer, d -> {
+                    List<OwnerReference> ownerReferences = d.getMetadata().getOwnerReferences();
+                    if (!ownerReferences.isEmpty()) {
+                        return Set.of(new ResourceID(ownerReferences.get(0).getName(), d.getMetadata().getNamespace()));
+                    } else {
+                        return Collections.emptySet();
+                    }
+        });
+
+        return List.of(deploymentEvent);
+    }
+
     @Override
     public UpdateControl<Keycloak> reconcile(Keycloak kc, Context context) {
-        logger.trace("Reconcile loop started");
-        final var spec = kc.getSpec();
-
-        logger.info("Reconciling Keycloak: " + kc.getMetadata().getName() + " in namespace: " + kc.getMetadata().getNamespace());
-
-        KeycloakStatus status = kc.getStatus();
-        var deployment = new KeycloakDeployment(client);
-
-        try {
-            var kcDeployment = deployment.getKeycloakDeployment(kc);
-
-            if (kcDeployment == null) {
-                // Need to create the deployment
-                deployment.createKeycloakDeployment(kc);
-            }
-
-            var nextStatus = deployment.getNextStatus(spec, status, kcDeployment);
-
-            if (!nextStatus.equals(status)) {
-                logger.trace("Updating the status");
-                kc.setStatus(nextStatus);
-                return UpdateControl.updateStatus(kc);
-            } else {
-                logger.trace("Nothing to do");
-                return UpdateControl.noUpdate();
-            }
-        } catch (Exception e) {
-            logger.error("Error reconciling", e);
-            status = new KeycloakStatus();
-            status.setMessage("Error performing operations:\n" + e.getMessage());
-            status.setState(KeycloakStatus.State.ERROR);
-            status.setError(true);
+        String kcName = kc.getMetadata().getName();
+        String namespace = kc.getMetadata().getNamespace();
+
+        Log.infof("--- Reconciling Keycloak: %s in namespace: %s", kcName, namespace);
 
+        var statusBuilder = new KeycloakStatusBuilder();
+
+        // TODO use caches in secondary resources; this is a workaround for https://github.com/java-operator-sdk/java-operator-sdk/issues/830
+        // KeycloakDeployment deployment = new KeycloakDeployment(client, config, kc, context.getSecondaryResource(Deployment.class).orElse(null));
+        var kcDeployment = new KeycloakDeployment(client, config, kc, null);
+        kcDeployment.updateStatus(statusBuilder);
+        kcDeployment.createOrUpdateReconciled();
```

Comment:
I might be missing something, are we recreating the deployment all of the times?

Reply:
Nope, resource version is not changed most of the times this is called. As far as I understand it, it's something like `kubectl apply`. Not sure if we like [`edit`](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#deployment) more. ;)

Reply:
I do not like `edit` more ...

But according to this [comment](https://github.com/fabric8io/kubernetes-client/issues/2454#issuecomment-708316924) the resource should always be updated.

Reply:
I'm not saying there's no API request with each call but if the resource is not effectively updated, its version remains the same. That's fine from my PoV, at least for now.

Reply:
I was looking for a method like `boolean shouldReconcile()` in `KeycloakDeployment`, but we can always add it later.
If you tested this a bit, I'm fine 👍

Reply:
Sure, let's keep it simple for the prototype for now and watch for possible issues. We can always add a complex logic for comparing and evaluating the need of calling `createOrUpdate`.

Reply:
Deal 👍